### PR TITLE
[SPARK-33004][SQL] Migrate DESCRIBE column to use UnresolvedTableOrView to resolve the identifier 

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

This PR proposes to migrate `DESCRIBE tbl colname` to use `UnresolvedTableOrView` to resolve the table/view identifier. This allows consistent resolution rules (temp view first, etc.) to be applied for both v1/v2 commands. More info about the consistent resolution rule proposal can be found in [JIRA](https://issues.apache.org/jira/browse/SPARK-29900) or [proposal doc](https://docs.google.com/document/d/1hvLjGA8y_W_hhilpngXVub1Ebv8RsMap986nENCFnrg/edit?usp=sharing).

### Why are the changes needed?

The current behavior is not consistent between v1 and v2 commands when resolving a temp view.
In v2, the `t` in the following example is resolved to a table:
```scala
sql("CREATE TABLE testcat.ns.t (id bigint) USING foo")
sql("CREATE TEMPORARY VIEW t AS SELECT 2 as i")
sql("USE testcat.ns")
sql("DESCRIBE t i") // 't' is resolved to testcat.ns.t

Describing columns is not supported for v2 tables.;
org.apache.spark.sql.AnalysisException: Describing columns is not supported for v2 tables.;
```
whereas in v1, the `t` is resolved to a temp view:
```scala
sql("CREATE DATABASE test")
sql("CREATE TABLE spark_catalog.test.t (id bigint) USING csv")
sql("CREATE TEMPORARY VIEW t AS SELECT 2 as i")
sql("USE spark_catalog.test")
sql("DESCRIBE t i").show // 't' is resolved to a temp view

+---------+----------+
|info_name|info_value|
+---------+----------+
| col_name|         i|
|data_type|       int|
|  comment|      NULL|
+---------+----------+
```

### Does this PR introduce _any_ user-facing change?

After this PR, `DESCRIBE t i` is resolved to a temp view `t` instead of `testcat.ns.t`.

### How was this patch tested?

Added a new test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 1]
```diff
@@ -168,7 +168,31 @@ class DataSourceV2SQLSuite
       Array("Provider", "foo", ""),
       Array(TableCatalog.PROP_OWNER.capitalize, defaultUser, ""),
       Array("Table Properties", "[bar=baz]", "")))
+  }
 
+  test("Describe column is not supported for v2 catalog") {
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:173]: There doesn't seem to be a test coverage for this, so I am adding here.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 2]
```diff
@@ -256,16 +256,20 @@ class ResolveSessionCatalog(
     case RenameTableStatement(TempViewOrV1Table(oldName), newName, isView) =>
       AlterTableRenameCommand(oldName.asTableIdentifier, newName.asTableIdentifier, isView)
 
-    case DescribeRelation(ResolvedTable(_, ident, _: V1Table), partitionSpec, isExtended) =>
+    case DescribeRelation(r @ ResolvedTable(_, ident, _: V1Table), partitionSpec, isExtended)
+        if isSessionCatalog(r.catalog) =>
```
Comment (imback82) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:260]: This change is for `DESCRIBE table`, but adding this check for consistency.

**Comment by imback82** [comment]
cc @cloud-fan 

Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:260]: hmm, I think only session catalog can return `V1Table`?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 3]
```diff
@@ -256,16 +256,20 @@ class ResolveSessionCatalog(
...
 
     // Use v1 command to describe (temp) view, as v2 catalog doesn't support view yet.
     case DescribeRelation(ResolvedView(ident), partitionSpec, isExtended) =>
       DescribeTableCommand(ident.asTableIdentifier, partitionSpec, isExtended)
 
-    case DescribeColumnStatement(tbl, colNameParts, isExtended) =>
-      val name = parseTempViewOrV1Table(tbl, "Describing columns")
-      DescribeColumnCommand(name.asTableIdentifier, colNameParts, isExtended)
+    case DescribeColumn(r @ ResolvedTable(_, _, _: V1Table), colNameParts, isExtended)
+        if isSessionCatalog(r.catalog) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:268]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 4]
```diff
@@ -168,7 +168,31 @@ class DataSourceV2SQLSuite
...
+    withTable("testcat.tbl") {
+      spark.sql("CREATE TABLE testcat.tbl (id bigint) USING foo")
+      val ex = intercept[AnalysisException] {
+        spark.sql("DESCRIBE testcat.tbl id")
+      }
+      assert(ex.message.contains("Describing columns is not supported for v2 tables"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:179]: If there is no API limitation to support it, can we implement describe column for v2 catalogs later? @imback82

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala [hunk 5]
```diff
@@ -315,6 +315,17 @@ case class DescribeRelation(
   override def output: Seq[Attribute] = DescribeTableSchema.describeTableAttributes()
 }
 
+/**
+ * The logical plan of the DESCRIBE relation_name col_name command that works for v2 tables.
+ */
+case class DescribeColumn(
+    relation: LogicalPlan,
+    colNameParts: Seq[String],
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: is it possible that the column name is qualified like `catalog_name.ns1.tbl1.col`? Does Hive support qualified column names in `DESCRIBE`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: A simple idea is to put an `UnresolvedAttribute` here, and analyzer can do the work for us.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 2 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:260]: OK. I will remove the check here and other existing places.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 4 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:179]: OK, I will take a look.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala (hunk 5 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: > Does Hive support qualified column names in `DESCRIBE`?

Hive [doesn't support qualified name](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Describe) for column names. For example:

```
DESCRIBE hivesampletable default.hivesampletable.clientid

java.sql.SQLException: Error while processing statement: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.DDLTask. cannot find field default from [0:clientid, 1:querytime, 2:market, 3:deviceplatform, 4:devicemake, 5:devicemodel, 6:state, 7:country, 8:querydwelltime, 9:sessionid, 10:sessionpagevieworder]
```

Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: > A simple idea is to put an `UnresolvedAttribute` here, and analyzer can do the work for us.

Since we need to have the relation resolved first, we need to match like the following in the analyzer:
```scala
case DescribeColumn(r: ResolvedTable, u: UnresolvedAttribute, _) =>
...
```
Is that what you had in mind?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/DescribeTableSchema.scala [hunk 6]
```diff
@@ -18,7 +18,7 @@
 package org.apache.spark.sql.catalyst.plans
 
 import org.apache.spark.sql.catalyst.expressions.AttributeReference
-import org.apache.spark.sql.types.{MetadataBuilder, StringType, StructField, StructType}
+import org.apache.spark.sql.types.{MetadataBuilder, StringType}
 
 private[sql] object DescribeTableSchema {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/DescribeTableSchema.scala:23]: we should rename it to `DescribeCommandSchema`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: If Hive doesn't support it, it's OK to follow Hive here. Let's make sure that other commands only allow column name without qualifies too.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 7]
```diff
@@ -256,16 +256,18 @@ class ResolveSessionCatalog(
     case RenameTableStatement(TempViewOrV1Table(oldName), newName, isView) =>
       AlterTableRenameCommand(oldName.asTableIdentifier, newName.asTableIdentifier, isView)
 
-    case DescribeRelation(ResolvedTable(_, ident, _: V1Table), partitionSpec, isExtended) =>
+    case DescribeRelation(r @ ResolvedTable(_, ident, _: V1Table), partitionSpec, isExtended) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:259]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 8]
```diff
@@ -256,16 +256,18 @@ class ResolveSessionCatalog(
...
       DescribeTableCommand(ident.asTableIdentifier, partitionSpec, isExtended)
 
     // Use v1 command to describe (temp) view, as v2 catalog doesn't support view yet.
     case DescribeRelation(ResolvedView(ident), partitionSpec, isExtended) =>
       DescribeTableCommand(ident.asTableIdentifier, partitionSpec, isExtended)
 
-    case DescribeColumnStatement(tbl, colNameParts, isExtended) =>
-      val name = parseTempViewOrV1Table(tbl, "Describing columns")
-      DescribeColumnCommand(name.asTableIdentifier, colNameParts, isExtended)
+    case DescribeColumn(r @ ResolvedTable(_, _, _: V1Table), colNameParts, isExtended) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:266]: nit: `ResolvedTable(_, ident, _: V1Table)`

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 9]
```diff
@@ -318,7 +320,7 @@ class ResolveSessionCatalog(
           ignoreIfExists = c.ifNotExists)
       }
 
-    case RefreshTable(r @ ResolvedTable(_, _, _: V1Table)) if isSessionCatalog(r.catalog) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:321]: Ah sorry I missed these existing checks. If we must change some places to make it consistent, I think it's better to always add this `isSessionCatalog` check.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala (hunk 5 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: Sure, will take a look.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala (hunk 5 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: > is it possible that the column name is qualified like `catalog_name.ns1.tbl1.col`? Does Hive support qualified column names in `DESCRIBE`?

@cloud-fan actually I didn't respond to your first question. Yes, Spark allows qualifying the column name (same in 2.4):
```
scala> sql("CREATE TABLE t (id bigint) USING csv")
scala> sql("describe t spark_catalog.default.t.id").show
+---------+----------+
|info_name|info_value|
+---------+----------+
| col_name|        id|
|data_type|    bigint|
|  comment|      NULL|
+---------+----------+
```
, whereas Hive does not support. I guess we cannot change the behavior at this point?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:323]: Yea we can't break it. When we add v2 `DecribeColumExec` later, we may want to change `Seq[String]` to `Attribute` here, so that we can unify the column resolution logic for v1 and  v2 commands.
